cpp/src/arrow/compute/exec/vector_hash_test.cc

```diff
@@ -0,0 +1,287 @@
+  std::vector<uint8_t> unique_keys;
+  unique_keys.resize(num_bytes);
+  std::set<std::string> unique_key_strings;
+  for (int i = 0; i < num_unique; ++i) {
+    for (;;) {
+      int next_length;
+      if (use_varlen_input) {
+        next_length = min_length + random.next() % (max_length - min_length + 1);
+      } else {
+        next_length = max_length;
+      }
+      unique_keys_offsets[i + 1] = unique_keys_offsets[i] + next_length;
+      unique_keys.resize(unique_keys_offsets[i + 1]);
+      uint8_t* next_key = unique_keys.data() + unique_keys_offsets[i];
+
+      for (int iword = 0; iword < next_length / static_cast<int>(sizeof(uint64_t));
+           ++iword) {
+        reinterpret_cast<uint64_t*>(next_key)[iword] = random.next();
+      }
+      if (next_length % sizeof(uint64_t) > 0) {
+        uint8_t* tail = next_key + next_length - (next_length % sizeof(uint64_t));
+        for (int ibyte = 0; ibyte < (next_length % static_cast<int>(sizeof(uint64_t)));
+             ++ibyte) {
+          tail[ibyte] = static_cast<uint8_t>(random.next() & 0xff);
+        }
+      }
+      std::string next_key_string =
+          std::string(reinterpret_cast<const char*>(next_key), next_length);
+      if (unique_key_strings.find(next_key_string) == unique_key_strings.end()) {
+        unique_key_strings.insert(next_key_string);
+        break;
+      }
+    }
+  }
+
+  std::vector<int> row_ids;
+  row_ids.resize(num_rows);
+  std::vector<uint8_t> keys;
+  std::vector<uint32_t> keys_offsets;
+  keys_offsets.resize(num_rows + 1);
+  keys_offsets[0] = 0;
+  for (int i = 0; i < num_rows; ++i) {
+    int row_id = random.next() % num_unique;
+    row_ids[i] = row_id;
+    int next_length = unique_keys_offsets[row_id + 1] - unique_keys_offsets[row_id];
+    keys_offsets[i + 1] = keys_offsets[i] + next_length;
+  }
+  keys.resize(keys_offsets[num_rows]);
+  for (int i = 0; i < num_rows; ++i) {
+    int row_id = row_ids[i];
+    int next_length = keys_offsets[i + 1] - keys_offsets[i];
+    memcpy(keys.data() + keys_offsets[i],
+           unique_keys.data() + unique_keys_offsets[row_id], next_length);
+  }
+
+  constexpr int min_rows_for_timing = 1 << 23;
+  int num_repeats = static_cast<int>(bit_util::CeilDiv(min_rows_for_timing, num_rows));
+#ifndef NDEBUG
+  num_repeats = 1;
+#endif
+
+  std::vector<uint32_t> hashes_scalar32;
+  std::vector<uint64_t> hashes_scalar64;
+  hashes_scalar32.resize(num_rows);
+  hashes_scalar64.resize(num_rows);
+  std::vector<uint32_t> hashes_simd32;
+  std::vector<uint64_t> hashes_simd64;
+  hashes_simd32.resize(num_rows);
+  hashes_simd64.resize(num_rows);
+
+  int64_t hardware_flags_scalar = 0LL;
+  int64_t hardware_flags_simd = ::arrow::internal::CpuInfo::AVX2;
+
+  constexpr int mini_batch_size = 1024;
+  std::vector<uint32_t> temp_buffer;
+  temp_buffer.resize(mini_batch_size * 4);
+
+  printf("cycles per hash ");
+  for (bool use_simd : {false, true}) {
+    auto start = std::chrono::high_resolution_clock::now();
```

Comment:
Is it ok to do benchmarking inside of the unit test? Maybe we should separate this out into its own benchmark?

Reply:
You are right. I removed the benchmarking parts from tests.